### DISCUSSION TITLE
rd2markdown collaboration

### DISCUSSION THREAD

**Discussion by dgkf-roche** [quantsch/Rd2md#27] [General]
Hi @quantsch 👋  

I'm reaching out on behalf of [Genentech/rd2markdown](https://github.com/Genentech/rd2markdown), developed primarily by myself and @maksymiuks. 

We started development on `rd2markdown` a couple years back. `Rd2md` existed that time, but development looked like it was a bit sporadic and was lacking a few key features that we needed urgently so we whipped up something that suited our needs at the time. At that time, `Rd2md` used a collection of `if-else` blocks to handle Rd tags and that structure didn't quite live up to the level of extensibility that we needed. 

Fast forward to today, and we're preparing to present on `rd2markdown` at the upcoming UseR! We wanted to check in with this project before sharing our progress. When checking in on the status, we were very happy to see that development seems to be more constant and that you have undertaken quite an impressive refactor exercise that really starts to make the packages look very similar architecturally to ours!

Seeing that both packages have similar goals, and more recently even have similar design, we wanted to reach out to float the idea of merging efforts.

**Comment by quantsch** [comment]
Hi @dgkf-roche

Thanks for reaching out. Great to hear that you are presenting at UseR!

I just enabled "Discussions" on the repo and migrated your issue here.

A bit of history: when I started this repo, it was actually a fork of another repo which did the implementations with these `if-esle` blocks and just allowed for parsing single Rd.files. My contribution was just to implement a quick wrapper function to combine the singel Rd files into the markdown reference manual as one document. There was a specific use case for me back then for this ref man markdown. But I always wanted to come back and start from scratch, which I did earlier this year.

Had a very quick look at your repo and indeed looks the approach is fairly similar - very nice work! I like your test coverage :)

Happy to have a deeper look and gather some thoughts about how we could merge efforts.

**Reply by dgkf-roche** [reply]
Wonderful. Take your time. Ours isn't up on CRAN yet, and we delayed publishing ours after seeing that activity on this project has been more constant.

Just to give you a brief history of our project and goals:

- We started it a couple years ago to support our package validation pipelines. We work at a pharmaceutical company where documenting the behavior of software used in clinical analysis is a regulatory requirement. We render Rd's to md's as one part of a comprehensive report.
- Since then, our package has been used to generate ~3,000 reports
- Our priority has been to generate markdown output for specific fragments of Rds for inclusion in markdown templates, used to render to pdf. 

Brief comparison:

- Yours is already on CRAN, and we want to avoid muddying the space with duplicitous packages
- Personally, I think `Rd2md` is the better package name, following the trend of `tools::Rd2*` functions and `as_markdown()` is a better generic name than `rd2markdown` - especially because we often operate on things other than `rd`'s like tags.
- I think you've covered more tags than we have.
- Our output has excess newlines, which it looks like you also had to deal with in `trim_ws_nodes`. We've ventured down this path too. In our implementation, I've considered emitting output to a shared `textConnection` buffer, which might avoid having multiple passes of cleanup.
- Your package is licensed under the GPL, while ours is MIT, meaning that we could port code over from ours, but you'd need to relicense if we were to converge on our codebase. 

So in short, if you'd be open to it, I think your package might provide the better foundation for a collaboration. On our end, we'd like to have some level of confidence that if we hit bugs that prevent us using the package in a production setting, that we'd have the ability to introduce changes or expect reasonably quick turnaround on PRs. We can always fork for hotfixes if that's not an option, but would prefer a more direct line of contributing.

**Reply by quantsch** [reply]
Sorry for the late reply, I didn't get around to having a deeper look until now.

I think the general ideas of the implementations are similar enough that merged efforts make sense.

Some notes:
* Testing: In order to use Rd2md in a production context, we definitively need to prioritize writing/porting tests, so you can ensure stability on your side. Given you have already put significant effort into testing, with some effort this should be doable.
* Seems you require the selection of individual `fragments`. We would need to port this feature, which should be fairly easy.
* Paragraphs/newlines: It's not really working exactly as expected, see e.g. https://github.com/quantsch/Rd2md/issues/23. I want to go over the whole flattening functions again anyway and refactor it, ideally simplify it as I personally find the code hard to read as-is.
* Do you actually use custom macros in your package documentations? If yes, we would need to port this feature https://github.com/Genentech/rd2markdown/blob/main/R/get_rd.R#L38
* The output of `documentation_to_markdown()` is a list of markdowns of the individual files. This can be easily integrated as an additional exported function.
* With more exported functions, we will need to think of a good, systematic way of naming them (since naming things is one of the two hard problems in computer science :) )
* I am happy to provide you with the necessary rights so you can apply fixes in case of problems autonomously. Do your pipelines pull the package from CRAN or Github?

**Reply by dgkf-roche** [reply]
Wonderful - glad to hear that you see value in combining efforts as well. 

> Paragraphs/newlines: It's not really working exactly as expected

No worries - ours isn't either 😂 . We currently handle ours much the same way you do - through a post-processing cleanup that tries to correct for some of the issues that can come up when treating each tag as independent. I've been kicking around some ideas for improvement, but we haven't committed to anything yet. Happy to add to your discussion if this is an area of focus for you too.

> Do you actually use custom macros in your package documentations?

In our use case we're more often rendering documentation of other peoples' packages. Even if we don't handle the macros when rendering them to md, we still need `tools:::processRdSexprs` to know about them. 

> Do you actually use custom macros in your package documentations?

Ideally we pull packages from an in-house CRAN mirror, but we have a few edge cases that we pull from GitHub. I don't think we'd have any trouble managing a remote if needed. We can always lock to a commit/tag or keep a fork as a mirror.
